### ISSUE TITLE
DelaunayTriangulator is occasionally incorrect (6-point example)

### ISSUE THREAD

**Issue by SpexGuy** [libgdx/libgdx#4489]
#### Issue details
I am attempting to use the `DelaunayTriangulator` to calculate adjacency information in a voronoi diagram.  Unfortunately, it appears that the triangulator occasionally triangulates quads in the wrong direction.  Here's a picture of one of the failing cases I'm encountering:

![delaunay_fail](https://cloud.githubusercontent.com/assets/2281818/21246526/f6f9abb0-c2ee-11e6-863f-5d43f53d1001.png)

The background is the rasterized voronoi of the points, to emphasize the problem.  The quad made up of points `(0, 3, 1, 2)` should be triangulated into `(0, 3, 1)` and `(1, 2, 0)`, but instead it gets cut in the other direction.  Adding another point somewhere else in the diagram (even far away) can sometimes flip this particular quad to be correct.

#### Reproduction steps/code
This test case reproduces the issue exactly as shown in the picture.
```java
import com.badlogic.gdx.math.DelaunayTriangulator;
import com.badlogic.gdx.utils.ShortArray;

public class DelaunayTest {

	public static void main(String[] args) {
		float[] points = new float[12];
		points[0] = (float) 2000.7620849609375;
		points[1] = (float) 3044.865234375;
		points[2] = (float) 3121.152099609375;
		points[3] = (float) 2331.909912109375;
		points[4] = (float) 1909.64794921875;
		points[5] = (float) 1924.1240234375;
		points[6] = (float) 3096.233642578125;
		points[7] = (float) 2911.776611328125;
		points[8] = (float) 776.0308227539062;
		points[9] = (float) 658.64501953125;
		points[10] = (float) 3619.53564453125;
		points[11] = (float) 163.5050811767578;

		DelaunayTriangulator delaunay = new DelaunayTriangulator();
		ShortArray trigs = delaunay.computeTriangles(points, false);

		for (int c = 0, n = trigs.size; c < n; c += 3) {
			System.out.format("(%d, %d, %d)\n", trigs.get(c), trigs.get(c+1), trigs.get(c+2));
		}

		// Prints out:
		// (4, 2, 5)
		// (2, 3, 1) // wrong!
		// (2, 4, 0)
		// (3, 5, 1)
		// (2, 0, 3) // wrong!
		// (5, 2, 1)
	}

}
```

#### Version of LibGDX and/or relevant dependencies
master (as of 12/15/2016, commit 52bbab3)

#### Please select the affected platforms
- [ ] Android
- [ ] iOS (robovm)
- [ ] iOS (MOE)
- [ ] HTML/GWT
- [X] Windows
- [ ] Linux
- [ ] MacOS
I've only tested on Windows 7, but the problem seems algorithmic.

**Comment by SpexGuy** [comment]
Here's the above example, with the circumcircles drawn on it.  The point highlighted in red is well inside the circumcircle of the triangle formed by the points circled in blue.  The circumCircle(..) function in DelaunayTriangulator.java correctly identifies it as such. It also identifies the opposite point as being inside the triangle on the other side of the offending edge.
![delaunay_fail_circles](https://cloud.githubusercontent.com/assets/2281818/21270184/24c32378-c37b-11e6-8d9c-456ff5487c25.png)

**Comment by slebed** [comment]
I can confirm this issue as well.  I'm using the DelaunayTriangulator for creating dynamic navmeshes and adding points outside an incorrect region will fix the error by flipping an edge.

**Comment by SpexGuy** [comment]
All I actually needed was to check adjacency of two points, not calculate the entire triangulation.  So I made this class to do that:
https://gist.github.com/SpexGuy/3ef9ed3561fb1f2c32a455ec8c0ccef3

I'm leaving the issue open, because it is actually a bug, but I no longer need a fix.

**Comment by Heerbann** [comment]
why not send a pull request if you have fixed it?

**Comment by SpexGuy** [comment]
@Heerbann I didn't fix it.  I solved the simpler, specific problem I had so that I didn't need to do a full delaunay triangulation and hence didn't need LibGDX's DelaunayTriangulator.  I could add my class to LibGDX with a pull request, but it isn't very useful unless you only need to determine whether pairs of points are adjacent in the triangulation and you have an efficient query to find points in a (possibly extremely large) circle.

**Comment by tommyettinger** [comment]
So we contributors have been going through issues recently and closing any that no longer represent legitimate bugs. This is not one of those cases. The example code provided by @SpexGuy was very helpful in confirming the bug, but actually solving it was a can of worms, and I'm not certain I've got a fix for all buggy situations... But it's something. What I found was that the point sorting code was sometimes not producing a correct ascending-x order, and while I'm not totally sure why this happens, an extra check is all that's needed to make sure that the sort is correct. Once the sort is correct, the triangles match the diagram given in the first post. This is a two-line fix, and as long as there isn't some extra unknown bug, it should be enough. I'll PR the change in a few minutes.
